src/Chainweb/Pact/PactService/Checkpointer.hs

```diff
@@ -377,23 +380,38 @@ rewindTo rewindLimit (Just (ParentHeader parent)) = do
                             (\(!p) (!c) -> runPact (fastForward (ParentHeader p, c)) >> return c)
                             (return h) -- initial parent
                             return
+                        & progress runPact progressLogger 25000
                         & S.length_
             logInfo $ "rewindTo.playFork: replayed " <> sshow c <> " blocks"
 
     -- This provides some progress feedback in case of long pact replays
     -- (e.g. when revalidating the pact history)
-    progress block = do
-        let h = _blockHeight block
-        when (h `rem` 10000 == 0) $
-            logInfo $ "rewindTo.fastForward: replay block at height " <> sshow (_blockHeight block)
+    --
+    progress
+        :: (PactServiceM cas () -> IO ())
+            -- ^ Callback for running a pact action (used for logging)
+        -> P.Logger
+            -- ^ logger for loggingprogress
+        -> Int
+            -- ^ How often a progress message is issued
+        -> S.Stream (S.Of BlockHeader) IO r
+        -> S.Stream (S.Of BlockHeader) IO r
+    progress runPact logger (n :: Int) s = s
+        & S.zip (S.enumFrom 1)
+        & S.mapM
+            (\(!i, !b) -> do
```

Comment:
Sorry, I don't understand `Stream` constructions, that is, `S.Stream a IO r -> S.Stream a IO r` ...

- does `progress` act as an independent streaming consumer alongside the `S.scanM` call above?
- The `S.mapM` call here naively looks like "every time progress gets called it loops through stuff" but is it instead the incremental consumer?

Reply:
Composing streams fuses them into a single loop.

Reply:
That's a nice feature of the streaming library the stream are composed just by function composition.